src/morse_code_decoderii/codewars/m_sulkouski/github/com/MorseCodeDecoder_II.java

```diff
@@ -0,0 +1,179 @@
+                currentBitCounter++;
+                index++;
+            }
+
+            for (int i = 0; i < currentBitCounter / switchSpeed; i++) {
+                processedBits.append(currentBitValue);
+            }
+        }
+
+        index = 0;
+
+        while (index < processedBits.length()) {
+            currentBitCounter = 0;
+            currentBitValue = processedBits.charAt(index);
+            while (index < processedBits.length() && processedBits.charAt(index) == currentBitValue) {
+                currentBitCounter++;
+                index++;
+            }
+            if (currentBitValue == '1') {
+                switch (currentBitCounter) {
+                    case 1:
+                        morseCode.append(".");
+                        break;
+                    case 3:
+                        morseCode.append("-");
+                        break;
+                    default:
+                }
+            } else {
+                switch (currentBitCounter) {
+                    case 3:
+                        morseCode.append(" ");
+                        break;
+                    case 7:
+                        morseCode.append("   ");
+                        break;
+                    default:
+                }
+            }
+        }
+
+        return morseCode.toString();
+    }
+
+    private static String trimBits(String rawBits) {
+        StringBuilder str = new StringBuilder(rawBits);
+        int index = str.length() - 1;
+        while (index >=0 && str.charAt(index) == '0') {
+            str.deleteCharAt(index);
+            index--;
+        }
+
+        if (str.length() == 0) {
+            return "";
+        }
+
+        index = 0;
+        while(str.charAt(index) == '0') {
+            str.deleteCharAt(index);
+        }
+
+        return str.toString();
+    }
+    private static int detectSwitchSpeed(String bits) {                             //switch speed, i.e. bit ratio speed, equals the smallest number of 0's or 1's in a
+        if (bits.length() <= 1) {                                                   //transmission. if >= 0, switch speed is 1
+            return 1;
+        }
+        int index = 0;
+        int endIndex = bits.length() - 1;
+        int minBitCounter = endIndex;
+        int currentBitCounter = 0;
+        char currentValue;
+
+        while (index < endIndex) {
+            currentValue = bits.charAt(index);
+
+            while (index <= endIndex && currentValue == bits.charAt(index)) {
+                index++;
+                currentBitCounter++;
+            }
+
+            if (currentBitCounter < minBitCounter) {
+                minBitCounter = currentBitCounter;
+            }
+            currentBitCounter = 0;
+        }
+        return minBitCounter;
+    }
+
+    private static String decodeMorse(String codedText) {
+        codedText = decodeBits(codedText);
+        String[] codedPhrase = codedText.split("   ");
```

Comment:
for example, the string could be a constant of MourseCode class and be reused in many places.